internal/app/referenceclient/impl.go

```diff
@@ -117,54 +117,10 @@ func (i *invoker) unary(
 	ctx context.Context,
 	req *conformancev1.ClientCompatRequest,
 ) (*conformancev1.ClientResponseResult, error) {
-	msg := req.RequestMessages[0]
-	ur := &conformancev1.UnaryRequest{}
-	if err := msg.UnmarshalTo(ur); err != nil {
-		return nil, err
-	}
-
-	request := connect.NewRequest(ur)
-
-	// Add the specified request headers to the request
-	internal.AddHeaders(req.RequestHeaders, request.Header())
-
-	var protoErr *conformancev1.Error
-	var headers []*conformancev1.Header
-	var trailers []*conformancev1.Header
-	payloads := make([]*conformancev1.ConformancePayload, 0, 1)
-
-	ctx = i.withWireCapture(ctx)
-
-	// Invoke the Unary call
-	resp, err := i.client.Unary(ctx, request)
-
-	if err != nil {
-		// If an error was returned, first convert it to a Connect error
-		// so that we can get the trailers from the Meta property. Then,
-		// convert _that_ to a proto Error so we can set it in the response.
-		connectErr := internal.ConvertErrorToConnectError(err)
-		trailers = internal.ConvertToProtoHeader(connectErr.Meta())
-		protoErr = internal.ConvertConnectToProtoError(connectErr)
-	} else {
-		// If the call was successful, get the headers and trailers
-		headers = internal.ConvertToProtoHeader(resp.Header())
-		trailers = internal.ConvertToProtoHeader(resp.Trailer())
-		// If there's a payload, add that to the response also
-		if resp.Msg.Payload != nil {
-			payloads = append(payloads, resp.Msg.Payload)
-		}
-	}
-
-	statusCode, feedback := i.examineWireDetails(ctx)
-
-	return &conformancev1.ClientResponseResult{
-		ResponseHeaders:  headers,
-		ResponseTrailers: trailers,
-		Payloads:         payloads,
-		Error:            protoErr,
-		HttpStatusCode:   statusCode,
-		Feedback:         feedback,
-	}, nil
+	return doUnary(ctx, req, i, i.client.Unary,
+		func(resp *conformancev1.UnaryResponse) *conformancev1.ConformancePayload {
+			return resp.Payload
+		})
```

Comment:
I was making some small changes in here and realized that not only was this logic duplicated, but it was also subtly different between the two already. So I decided to address the TODO below to consolidate the logic.